### ISSUE TITLE
[QST] - Multi-GPU Support w/ Naive model.fit()

### ISSUE THREAD

**Issue by jacobdineen** [NVIDIA-Merlin/Transformers4Rec#423]
# ❓ Questions & Help

## Details

Hello! My team is using T4Rec w/ a BinaryClassificationTask Head. We are utilizing the recommended class method `model.fit()` rather than the trainer class. 

We are currently running experiments on various AWS instances (4xV100 / 8xV100) & it appears that the model is training on a single gpu. 

Should the naive `model.fit()` have distributed training out of the box? 

Thanks in advance!

CLI output below during training:
```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 470.57.02    Driver Version: 470.57.02    CUDA Version: 11.4     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla V100-SXM2...  On   | 00000000:00:1B.0 Off |                    0 |
| N/A   62C    P0    88W / 300W |   2132MiB / 16160MiB |      7%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   1  Tesla V100-SXM2...  On   | 00000000:00:1C.0 Off |                    0 |
| N/A   45C    P0    39W / 300W |      3MiB / 16160MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   2  Tesla V100-SXM2...  On   | 00000000:00:1D.0 Off |                    0 |
| N/A   42C    P0    44W / 300W |      3MiB / 16160MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   3  Tesla V100-SXM2...  On   | 00000000:00:1E.0 Off |                    0 |
| N/A   45C    P0    40W / 300W |      3MiB / 16160MiB |      0%      Default |
|                               |                      |                  N/A
```

**Comment by gabrielspmoreira** [comment]
Hi @jacobdineen. We have currently two ways of training models with Transformers4Rec PyTorch API:

1. `Trainer.train()`
2. `Model.fit()`

The 2nd approach was created to make the PyTorch API aligns with the TF API (like Keras `model.fit()`). But for PyTorch API it is just a simple train loop, as you can see [here](https://github.com/NVIDIA-Merlin/Transformers4Rec/blob/083684baade456250dad6187f3e6a6cfb0659841/transformers4rec/torch/model/base.py#L592-L637), which does not support multi-GPU OOTB.

The `trainer.train()` is the most comprehensive training method, as we inherit our `Trainer` from the HuggingFace's Trainer and use its [train()](https://github.com/huggingface/transformers/blob/v4.12.5/src/transformers/trainer.py#L965) method, which leverages Torch's `DataParallel` and `DistributedDataParallel` options for multi-GPU training, as mentioned in this [doc](https://nvidia-merlin.github.io/Transformers4Rec/main/training_eval.html).

So for multi-GPU data parallel training I suggest either using our `Trainer.train()` or creating you custom train loop based on that method, wrapping the model with `DataParallel` or `DistributedDataParallel`.

**Comment by jacobdineen** [comment]
Thank you, @gabrielspmoreira!
You can close this issue. My team and I will try to write a custom train loop for the keras-style PyTorch API.

**Comment by jacobdineen** [comment]
@gabrielspmoreira Is there any chance that this will be added as a feature in later releases?

**Comment by gabrielspmoreira** [comment]
Hi @jacobdineen . Can you be a little bit more specific on what you are looking for. Do you need multi-GPU training but prefer a simpler training loop (like we have for `Model.fit()`) rather than using the Trainer class we inherit from HF Trainer?

**Comment by jacobdineen** [comment]
Apologies for not being clear. I'll send over some additional context tomorrow.

**Comment by rnyak** [comment]
@jacobdineen thanks for more contexts. I guess your answer to Gabriel's question `Do you need multi-GPU training but prefer a simpler training loop (like we have for Model.fit()) rather than using the Trainer class we inherit from HF Trainer?` is a Yes, isnt it?

**Comment by jacobdineen** [comment]
@rnyak Correct, ideally the simpler training loop would support multi-gpu distributed training.

**Comment by rnyak** [comment]
> My team and I will try to write a custom train loop for the keras-style PyTorch API.

@jacobdineen what's the current status with your custom work? I understand,  `DistributedDataParallel` did not work. Did you test `DataParallel` and gain any training run time perf improvement with it ? Thanks.

**Comment by jacobdineen** [comment]
@rnyak 
A member of the team is going to test out the solutions from [this issue](https://github.com/NVIDIA-Merlin/Transformers4Rec/issues/456) in the next day or so. 

**As for what we have tried - Assume the following setup:**

```python
# type: transformers4rec.torch.model.base.Model
model = transformer_config.to_torch_model(input_module, prediction_task)
model = model.cuda()


train_path = glob.glob(os.path.join(output_dir, f"1/train.parquet"))
train_loader = next(iter(get_dataloader(schema, train_path)))
```

The workflow we currently follow is incremental (as the tutorials are):
```python
for time_index in range(start_time_window_index, final_time_window_index):
    # fetch paths and dataloaders here
   ...........
    model.fit(train_loader, 
              optimizer=optimizer, 
              num_epochs=num_epochs, 
              amp=False, 
              verbose = False)

   # custom model evaluate code here
```


The issue with `torch.nn.DataParallel` is that we can not simply wrap the model and call `fit`, e.g., :
```python
import torch.nn as nn
wrapped_model = nn.DataParallel(model)
wrapped_model.fit(data)
```
Throws an attribute error that the now `DataParallel` object has no attribute fit. As discussed above, the torch model class inherits from torch.nn.module, so wrapping the object in DataParallel works for evaluation as it invokes torch.nn.Module.__call__ on the forward pass

So something like this still works for flowing forward:
```python
train_loader = next(iter(get_dataloader(schema, train_path)))
x, y = data[0]

wrapped_model = nn.DataParallel(model)
wrapped_model(x) - computes probas
```

But to my understanding (correct me if I'm wrong), accessing the t4rec model's class methods via `wrapped_model.module.fit()` doesn't invoke `DataParallel` at all.

**Comment by jacobdineen** [comment]
[This](https://discuss.pytorch.org/t/dataparallel-doesnt-work-when-calling-model-module-some-attribute/47556/6) is a pretty good example of the issue

**Comment by rnyak** [comment]
> this issue

thanks for more context.

**Comment by rnyak** [comment]
> [This](https://discuss.pytorch.org/t/dataparallel-doesnt-work-when-calling-model-module-some-attribute/47556/6) is a pretty good example of the issue

Thanks for more context.

**Comment by gabrielspmoreira** [comment]
This issue is going to be addressed by task #487

**Comment by viswa-nvidia** [comment]
@rnyak , please close this issue after tagging the PRs.
